src/defrag.c

```diff
@@ -1065,48 +1129,64 @@ void activeDefragCycle(void) {
-                                                   &defragfns, &ctx);
-                /* When done scanning the keyspace dict, we scan the expire dict. */
-                if (!cursor)
-                    expires_cursor = kvstoreDictScanDefrag(db->expires, slot, expires_cursor,
-                                                           scanCallbackCountScanned,
-                                                           &defragfns, NULL);
+                /* Iterates over all defragmentation stages, starting from the last unfinished stage. */
+                int num_stages = sizeof(defrag_stages) / sizeof(defrag_stages[0]);
+                serverAssert(defrag_stage < num_stages);
+                for (int i = defrag_stage; i < num_stages; i++) {
+                    if (defrag_stage != i) break; /* previous stage not finished */
+                    if (defrag_stages[i].kvs) {
+                        defrag_cursor = kvstoreDictScanDefrag(defrag_stages[i].kvs, slot, defrag_cursor,
+                            defrag_stages[i].scanfn, &defragfns, defrag_stages[i].ctx);
+                    }
+
+                    /* Move to next stage if current stage completed */
+                    if (!defrag_cursor) defrag_stage++;
+                }
+
+                /* All stages completed */
+                if (defrag_stage == num_stages) defrag_stage = 0;
             }
-            if (!(cursor || expires_cursor)) {
+
+            slot_finished = (!defrag_stage && !defrag_cursor);
+            if (slot_finished) {
                 /* Move to the next slot only if regular and large item scanning has been completed. */
                 if (listLength(db->defrag_later) > 0) {
                     defrag_later_item_in_progress = 1;
                     continue;
                 }
-                slot = kvstoreGetNextNonEmptyDictIndex(db->keys, slot);
+                slot = getNextClosestNonEmptySlot(defrag_stages, sizeof(defrag_stages)/sizeof(defrag_stages[0]), slot);
                 defrag_later_item_in_progress = 0;
-                ctx.slot = slot;
             }
     
             /* Once in 16 scan iterations, 512 pointer reallocations. or 64 keys
              * (if we have a lot of pointers in one hash bucket or rehashing),
              * check if we reached the time limit.
              * But regardless, don't start a new db in this loop, this is because after
              * the last db we call defragOtherGlobals, which must be done in one cycle */
-            if ((!(cursor || expires_cursor) && slot == -1) ||
+            if ((slot_finished && slot == -1) ||
                 ++iterations > 16 ||
                 server.stat_active_defrag_hits - prev_defragged > 512 ||
                 server.stat_active_defrag_scanned - prev_scanned > 64)
             {
-                if (!cursor || ustime() > endtime) {
+                if (slot_finished || ustime() > endtime) {
```

Comment:
https://github.com/redis/redis/pull/13058#discussion_r1510027980
@oranagra the reason I don't do this is that here, if we only defrag one stage at a time, then we'll have to go into the defrag stages more than once.
.e.g keys, expires, pubsubshard all have 4 slots, before we only needed 4 traversals, now we need 12 (4 * 3 slots) (but not a loop, just the current stage).

Reply:
you mean that the code wants to avoid starting the defragOtherGlobals in an activeDefragCycle that also handled anything else (since we can't break defragOtherGlobals into fragments)?
i'm not sure i follow, or why is that a restriction for the "inner loop" to be slots or steps.

Reply:
what i'm confused about is at who point we leave this loop without timeout.
1. end of current slot? #11695
2. end of current db? https://github.com/redis/redis/commit/be1b4aa9aa76d15bb57b7159cb559a8b9dde5513#diff-5d1e7cac0f87f2558c5cd612adedea707ebdc59c67ccd94ced70b3ad35e9e99f

following is the time consume of defrag cluster tests:

this PR (leval at the end of current db):
```
< [ok]: Active defrag main dictionary: cluster (12678 ms)
< [ok]: Active defrag eval scripts: cluster (2595 ms)
< [ok]: Active defrag big keys: cluster (11035 ms)
```

this PR (leval at the end of current slot):
```
> [ok]: Active defrag main dictionary: cluster (33192 ms)
> [ok]: Active defrag eval scripts: cluster (3004 ms)
> [ok]: Active defrag big keys: cluster (30273 ms)
```

@oranagra not sure which one is better, one (end of db) may bring higher latency, and the other on takes more time to defrag.

Reply:
i prefer lower latency.
i mean:
1. we should respect the defrag effort config, and try to exit as soon as we notice we spend our time limit.
2. (with the exception of the defragOtherGlobals), we should continue to defrag more items if we didn't reach the time limit (i.e. don't break just because we finished one step, start the next one right away)

in that sense, i'm not sure why at all we need to check `slot_finished` or `all_stages_finished` before probing the time?
we already have some conditions there to avoid the time check system call too often.

p.s. if the tests are too slow, we can maybe increase the effort configs to speed them up.

am i missing something?

Reply:
> 1. we should respect the defrag effort config, and try to exit as soon as we notice we spend our time limit.
> 2. (with the exception of the defragOtherGlobals), we should continue to defrag more items if we didn't reach the time limit (i.e. don't break just because we finished one step, start the next one right away)

the current implementation strictly respects the defrag config, as long as it doesn't timeout, we wil 
continue defragmenting util the current database is completed, that is, `all_stages_finished` is 1.

the reason #11695 consumes more time than current PR is that it leaves when a certain slot ends, even
if it doesn't timeout, but i think this is a bug. 

> 
> in that sense, i'm not sure why at all we need to check `slot_finished` or `all_stages_finished` before probing the time? we already have some conditions there to avoid the time check system call too often.

i think `all_stages_finished` is necessary, if we finish all the stages in one defrag cycle, and
it doesn't timeout yes, we can leave right away instead of starting again, which is likely to
be impossble to find fragments.

Reply:
> the reason https://github.com/redis/redis/pull/11695 consumes more time than current PR is that it leaves when a certain slot ends, even
if it doesn't timeout, but i think this is a bug.

good

> i think all_stages_finished is necessary, if we finish all the stages in one defrag cycle, and
it doesn't timeout yes, we can leave right away instead of starting again, which is likely to
be impossble to find fragments.

right, and we did get rid of `slot_finished`.

so all is good now?

Reply:
let me check it again, still not quite confident.